syntax = "proto3";

enum DBConnectionType {
  USER = 0;
  TRANSIENT = 1;
  GLOBAL = 2;
}

enum DBMethod {
  INITIALIZE = 0;
  OPEN_CURSOR = 1;
  CLOSE_CURSOR = 2;
  EXECUTE = 3;
  EXECUTEMANY = 4;
  EXECUTESCRIPT = 5;
  ROLLBACK = 6;
  COMMIT = 7;
  CLOSE = 8;
}

enum DBError {
  MemoryError = 0;
  HostSystemError = 1;
  ProgrammingError = 2;
  ComputationError = 3;
  DBWriteError = 4;
  UnknownError = 5;
}

message DBWriteResult {
  int32 rowcount = 1;
  uint64 lastrowid = 2;
}

message DBWriterError {
  string message = 1;
  DBError name = 2;
}

// using TypedArray because oneof doesn't allow repeated as label for their members
// docs: https://protobuf.dev/programming-guides/proto3/#using-oneof
message TypedArray {
  repeated TypedData array = 1;
}

message TypedData {
  string type = 1;
  oneof data {
    bytes bytes = 2;
    TypedArray array = 3;
  }
}

message ZMQCallData {
  DBConnectionType connection_type = 1;
  string db_path = 2;
  optional string cursor_name = 3;
  DBMethod method = 4;
  repeated TypedData args = 5;
  map<string, TypedData> kwargs = 6;
}

message ZMQReturnData {
  optional DBWriteResult result = 1;
  optional DBWriterError error = 2;
}
